### PR TITLE
Add EventLogger

### PR DESCRIPTION
**Pull request by frsv**
Event logger is a core handler of the logging hook.

It is almost a full copy of the EventLogger in the HiveProtoLoggingHook with minimal adjustments

### CODE REVIEW

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+  private void reportRetryState(GenericRecord event, int retryCount, IOException writeException) {
+    if (retryCount < MAX_RETRIES) {
+      LOG.warn(
+          "Error writing proto message for query {}, eventType: {}, retryCount: {},"
+              + " error: {} ",
+          event.get("QueryId"),
+          event.get("EventType"),
+          retryCount,
+          writeException.getMessage(),
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:208]: It is unnecessary to log the exception message as a parameter (the `error: {}` part of the error message above), if the whole exception is logged anyway (the next parameter: `writeException`). The error message would be duplicated in the log. I would suggest to remove this parameter: `writeException.getMessage()` and adjust the error message accordingly.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+          writeException);
+      LOG.trace("Exception", writeException);
+    } else {
+      LOG.error(
+          "Error writing proto message for query {}, eventType: {}: ",
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:213]: I would suggest to remove the colon `:` at the end of the error message.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+ * Main handler of query events. Processes the captured events and writes resulting records to a
+ * file.
+ */
+public class EventLogger {
+  private static final Logger LOG = LoggerFactory.getLogger(EventLogger.class);
+
+  private static final int MAX_RETRIES = 2;
+  private static final int SHUTDOWN_WAIT_TIME = 5;
+  private static final int QUERY_EVENTS_QEUUE_DEFAULT_SIZE = 64;
+  private static final Duration DEFAULT_ROLLOVER_TIME_MILLISECONDS = Duration.of(600, SECONDS);
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:43]: 1. I would suggest renaming to `DEFAULT_ROLLOVER_DURATION` as the value is not related to milliseconds.
2. The `Duration.of(600, SECONDS)` can be simplified to `Duration.of(10, MINUTES)`.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+/**
+ * Main handler of query events. Processes the captured events and writes resulting records to a
+ * file.
+ */
+public class EventLogger {
+  private static final Logger LOG = LoggerFactory.getLogger(EventLogger.class);
+
+  private static final int MAX_RETRIES = 2;
+  private static final int SHUTDOWN_WAIT_TIME = 5;
+  private static final int QUERY_EVENTS_QEUUE_DEFAULT_SIZE = 64;
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:42]: Please fix the typo in `QUEUE`.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+/**
+ * Main handler of query events. Processes the captured events and writes resulting records to a
+ * file.
+ */
+public class EventLogger {
+  private static final Logger LOG = LoggerFactory.getLogger(EventLogger.class);
+
+  private static final int MAX_RETRIES = 2;
+  private static final int SHUTDOWN_WAIT_TIME = 5;
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:41]: I would recommend changing the type to `Duration` or append a time unit to the name.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    ThreadFactory threadFactory =
+        new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setNameFormat("Migration Assessment Hook Query Log Writer %d")
+            .build();
+    logWriter = new ScheduledThreadPoolExecutor(1, threadFactory);
+
+    long rolloverInterval =
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:98]: I would suggest to add a time unit to the variable name.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  }
+
+
+  public void shutdown() {
+    if (logWriter != null) {
+      logWriter.shutdown();
+      try {
+        logWriter.awaitTermination(SHUTDOWN_WAIT_TIME, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:235]: The simplest way to handle `InterruptedException` is to call `Thread.currentThread().interrupt()` before `LOG.warn`.

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLoggerTest.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    ImmutableList<FileStatus> files =
+        ImmutableList.copyOf(fs.listStatus(directories.get(0).getPath()));
+    assertThat(files).hasSize(1);
+
+    FSDataInputStream inputStream = fs.open(files.get(0).getPath());
+
+    DatumReader<GenericRecord> reader = new GenericDatumReader<>(QUERY_EVENT_SCHEMA);
+
+    try (DataFileStream<GenericRecord> dataFileReader = new DataFileStream<>(inputStream, reader)) {
+      ArrayList<GenericRecord> records = new ArrayList<>();
```
Comment (dawidxc) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLoggerTest.java:203]: I would suggest to change the type of the variable to `List<GenericRecord>` (an interface).

**Review by dawidxc — APPROVED** [review]

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java (hunk 6 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+      return new DatePartitionedLogger(new Path(baseDir), conf, QUERY_EVENT_SCHEMA, clock);
+    } catch (IOException e) {
+      LOG.error("Unable to initialize logger, logging disabled.", e);
+    }
+
+    return null;
+  }
+
+  private String constructFileName() {
+    return "dwhassessment_" + loggerId + "-" + logFileCount + ".avro";
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+  }
+
+  private void handleTick() {
+    try {
+      maybeRolloverWriterForDay();
+    } catch (IOException e) {
+      LOG.error("Got IOException while trying to rollover: ", e);
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java (hunk 2 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java (hunk 1 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+          event.get("QueryId"),
+          event.get("EventType"),
+          writeException);
+    }
+  }
+
+  private void waitForRetry(int retryCount) {
+    try {
+      // 0 seconds, for first retry assuming fs object was closed and open will fix it.
+      Thread.sleep(1000L * retryCount * retryCount);
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+    } catch (InterruptedException e1) {
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java (hunk 7 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:235]: Done

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLoggerTest.java (hunk 8 above)
Comment (frsv) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLoggerTest.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+  private void handleTick() {
+    try {
+      maybeRolloverWriterForDay();
+    } catch (IOException e) {
+      LOG.error("Got IOException while trying to rollover: ", e);
+    }
+  }
+
+  private void maybeRolloverWriterForDay() throws IOException {
+    if (writer == null || !logger.getNow().equals(writerDate)) {
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java:171]: Acknowledged. I will do this as a separate PR, as change will involve several branches and it's a pain.

**Pull request merged**